### PR TITLE
WP-366 - E.COM | As a reader when I read a Special Report I want to understand it is part of a wider story collection

### PR DESCRIPTION
**Pull request by yasgreen93**
https://jira.economist.com/browse/WP-366
This adds the functionality for the special report in blog posts. CSS has been put in fe-blogs incase design want to make minor changes in the future.

### CODE REVIEW

**Comment by umbobabo** [comment]
Pls review the implementation abstracting the feature from the concept of 'Special Report'.
What the component need to do is generate a generic list of siblings.
Also be sure that we have the opportunity to customise each Classnames to match BEM as I asked previously to @garySiladi  in this implementation:

https://github.com/economist-components/component-sections-card/blob/master/src/parts/section-card-link.js#L6

File: src/index.js [hunk 1]
```diff
@@ -199,8 +200,10 @@ export default class BlogPost extends React.Component {
   render() {
     /* eslint-enableable complexity */
     const flyTitle = this.props.flyTitle;
-    const specialReportList = this.props.specialReportList.entries;
-    const isSpecialReport = this.props.printSectionName === 'Special report';
+    const showSiblingArticlesList = this.props.showSiblingArticlesList;
+    const sectionName = this.props.printSectionName;
```
Comment (umbobabo) [inline src/index.js]: The article should have just have a generic concept of section, not printSection.

File: src/index.js [hunk 2]
```diff
@@ -199,8 +200,10 @@ export default class BlogPost extends React.Component {
...
+    const elementClassName = sectionName.replace(' ', '-').toLowerCase();
```
Comment (umbobabo) [inline src/index.js]: sectionName could have also special character on it. We probably need a more accurate slug-ification: See something like: https://www.npmjs.com/package/slug

File: src/index.js [hunk 3]
```diff
@@ -199,8 +200,10 @@ export default class BlogPost extends React.Component {
...
+    const siblingArticles = showSiblingArticlesList ? this.props.specialReportList.entries : null;
```
Comment (umbobabo) [inline src/index.js]: The blog shouldn't know ahything about Special Report.
This is why we talk about 'smart' and 'dumb' component.
I just need to receive a list of siblings.
Move this logic on the smart component.

File: src/index.js [hunk 4]
```diff
@@ -277,22 +280,27 @@ export default class BlogPost extends React.Component {
...
-    const nextArticleLink = isSpecialReport ? nextSpecialReportArticle(specialReportList, flyTitle) : null;
     // Sometimes there is no rubric provided from Drupal and so there aren't as many elements in the array
     // This is checking if it exists to put the content after the first paragraph.
-    const blogText = isSpecialReport && content[1].props.children[2] ? (
-      content[1].props.children[2].props.text
-    ) : content[0].props.children[2].props.text;
-    if (isSpecialReport && (blogText || (content && nextArticleLink))) {
-      blogText.splice(1, 0, specialReportSideList);
+    let blogText = null;
+    if (showSiblingArticlesList && content[1].props.children[2]) {
```
Comment (umbobabo) [inline src/index.js]: I don't like this part. Is really confused.

File: src/parts/blog-post-siblings-list.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+export function siblingList(
+  articlesList,
+  flyTitle,
+  elementClassName,
+  sectionName
+) {
+  /* eslint-disable quotes*/
+  return (
+    <div className={`blog-post__${ elementClassName }-aside`}>
+      <span className={`blog-post__side-flytitle`}>{flyTitle}</span>
```
Comment (umbobabo) [inline src/parts/blog-post-siblings-list.js]: Do not use string template if there are no vars. Single quote instead.

File: src/parts/blog-post-siblings-list.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+      <span className={`blog-post__side-text`}>More in this {sectionName.toLowerCase()}:</span>
```
Comment (umbobabo) [inline src/parts/blog-post-siblings-list.js]: Better having this test overridable and set the current value as default.

File: src/parts/blog-post-siblings-list.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+      </ul>
+    </div>
+  );
+  /* eslint-enable quotes */
+}
+
+export function nextSiblingArticle(siblingArticles, flyTitle, elementClassName) {
+  let nextArticleinList = null;
+  /* eslint-disable array-callback-return */
+  siblingArticles.map((article, index) => {
```
Comment (umbobabo) [inline src/parts/blog-post-siblings-list.js]: I think Lint should suggest to use a forEach here. That should make the validation pass.

File: src/parts/blog-post-siblings-list.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    if (flyTitle === article.flyTitle) {
+      nextArticleinList = siblingArticles[index + 1];
+    }
+  });
+  /* eslint-enable array-callback-return */
+  return nextArticleinList ? (
+    <div className={classnames(`blog-post__${ elementClassName }-next-article`)}>
+      <a className={classnames(`blog-post__${ elementClassName }-next-article-link`)} href={nextArticleinList.webURL}>
+        → {nextArticleinList.flyTitle}: {nextArticleinList.title}
```
Comment (umbobabo) [inline src/parts/blog-post-siblings-list.js]: Overridable via a prop will make changes less painful and the component more re-usable.

File: src/index.js [hunk 9]
```diff
@@ -188,7 +199,14 @@ export default class BlogPost extends React.Component {
     return sectionDateAuthor;
   }
 
+  /* eslint-disable complexity */
   render() {
+    /* eslint-enableable complexity */
+    const flyTitle = this.props.flyTitle;
+    const showSiblingArticlesList = this.props.showSiblingArticlesList;
+    const sectionName = this.props.sectionName;
```
Comment (varholak-peter) [inline src/index.js]: Can we use destructuring here?

File: src/index.js [hunk 10]
```diff
@@ -265,6 +281,32 @@ export default class BlogPost extends React.Component {
       </div>
     );
     const TitleComponent = this.props.TitleComponent;
+    const articleHeader = showSiblingArticlesList ? (
+      <span className={classnames(`blog-post__${ elementClassName }-header`)}>
```
Comment (garySiladi) [inline src/index.js]: do we need here the `classnames` package? we're not combining classnames here.

File: src/index.js [hunk 11]
```diff
@@ -188,7 +199,14 @@ export default class BlogPost extends React.Component {
     return sectionDateAuthor;
   }
 
+  /* eslint-disable complexity */
   render() {
+    /* eslint-enableable complexity */
```
Comment (yasgreen93) [inline src/index.js]: Well spotted!

**Comment by varholak-peter** [comment]
LGTP

File: src/index.js [hunk 12]
```diff
@@ -188,7 +199,13 @@ export default class BlogPost extends React.Component {
     return sectionDateAuthor;
   }
 
+  /* eslint-disable complexity */
   render() {
+    /* eslint-enable complexity */
+    const { flyTitle, showSiblingArticlesList, sectionName } = this.props;
+    const elementClassName = showSiblingArticlesList && sectionName ? slug(sectionName, { lower: true }) : null;
+    const siblingArticles = showSiblingArticlesList && this.props.issueSiblings ?
+      this.props.issueSiblings.entries : null;
```
Comment (umbobabo) [inline src/index.js]: This 'this.props.issueSiblings.entries' lock the component to a very specific data structure while the component should be dump about that and just expose a prop 'issueSiblings' that accept an array (or maybe an object with an array, if you need a title).

File: src/index.js [hunk 13]
```diff
@@ -188,7 +199,13 @@ export default class BlogPost extends React.Component {
     return sectionDateAuthor;
   }
 
+  /* eslint-disable complexity */
```
Comment (umbobabo) [inline src/index.js]: You shouldn't disable this. It's for good.
You should review this method and simplify it.

File: src/index.js [hunk 14]
```diff
@@ -265,6 +282,32 @@ export default class BlogPost extends React.Component {
...
+    const innerContentElements = showSiblingArticlesList ? (content.filter((contentElement) => {
+      const innerContent = contentElement.key === 'inner-content';
+      return innerContent;
+    })[0].props.children) : null;
+    const blogPostTextElements = showSiblingArticlesList ? (innerContentElements.filter((contentElement) => {
+      const blogPostText = contentElement.key === 'blog-post__text';
+      return blogPostText;
+    })[0].props.text) : null;
+    if (showSiblingArticlesList && (blogPostTextElements || (content && this.props.nextArticleLink))) {
+      blogPostTextElements.splice(1, 0, siblingArticlesList);
```
Comment (umbobabo) [inline src/index.js]: Let create a parameter (prop) to set the paragraph in which we want the siblingList to appear.

File: src/parts/blog-post-siblings-list.js [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+  flyTitle,
+  elementClassName,
+  sectionName,
+  sideText
+) {
+  if (!articlesList) {
+    return null;
+  }
+  const blogSideText = sideText ? sideText : `More in this ${ sectionName.toLowerCase() }:`;
+  /* eslint-disable quotes*/
```
Comment (umbobabo) [inline src/parts/blog-post-siblings-list.js]: Why? Can't you just fix the quotes?

File: src/index.js [hunk 16]
```diff
@@ -188,7 +199,13 @@ export default class BlogPost extends React.Component {
     return sectionDateAuthor;
   }
 
+  /* eslint-disable complexity */
   render() {
+    /* eslint-enable complexity */
+    const { flyTitle, showSiblingArticlesList, sectionName } = this.props;
+    const elementClassName = showSiblingArticlesList && sectionName ? slug(sectionName, { lower: true }) : null;
```
Comment (umbobabo) [inline src/index.js]: Classnames based on data like section name can be a problem when you have dynamic data as we use to have.
Better expose a property as you did for the siblingList.
In this way you can also remove the slug.
For i.e. we will use the ID on fe-blogs that is more reliable than the section name.

**Pull request merged**